client/auth.go

```diff
@@ -64,50 +74,95 @@ func (c *Conn) readInitialHandshake() error {
 		// mysql-proxy also use 12
 		// which is not documented but seems to work.
 		c.salt = append(c.salt, data[pos:pos+12]...)
+		pos += 13
+		// auth plugin
+		if end := bytes.IndexByte(data[pos:], 0x00); end != -1 {
+			c.authPluginName = string(data[pos : pos+end])
+		} else {
+			c.authPluginName = string(data[pos:])
+		}
+	}
+
+	// if server gives no default auth plugin name, use a client default
+	if c.authPluginName == "" {
+		c.authPluginName = defaultAuthPluginName
 	}
 
 	return nil
 }
 
+// generate auth response data according to auth plugin
+func (c *Conn) genAuthResponse(authData []byte) ([]byte, bool, error) {
```

Comment:
That's a good question. Sure, the bool determines whether to add a \NUL to the end of data. That is quite tricky actually because mysql dev doc is very unclear about the message endings. The rules here are interpreted from MySQL server response packets. Besides, the MySQL implementation does not quite follow their docs in some response messages.

see: https://dev.mysql.com/doc/internals/en/string.html